radix-transactions/src/manifest/static_manifest_interpreter.rs

```diff
@@ -0,0 +1,760 @@
+        visitor: &mut V,
+        source_amount: ProofSourceAmount<'a>,
+    ) -> ControlFlow<V::Error<'a>> {
+        match source_amount.proof_kind() {
+            ProofKind::BucketProof(bucket) => {
+                self.get_existing_bucket::<V>(bucket)?.proof_locks += 1;
+            }
+            ProofKind::AuthZoneProof => {}
+        }
+        let new_proof = ManifestProof(self.proof_state.len() as u32);
+        self.proof_state.push(ProofState {
+            name: self
+                .manifest
+                .get_known_object_names_ref()
+                .known_proof_name(new_proof),
+            created_at: self.location,
+            consumed_at: None,
+            source_amount,
+        });
+        visitor.on_new_proof(new_proof, source_amount)
+    }
+
+    fn get_existing_proof<V: ManifestInterpretationVisitor>(
+        &mut self,
+        proof: ManifestProof,
+    ) -> ControlFlow<V::Error<'a>, &mut ProofState<'a>> {
+        match self.proof_state.get_mut(proof.0 as usize) {
+            Some(state) => {
+                if state.consumed_at.is_some() {
+                    ControlFlow::Break(
+                        ManifestValidationError::ProofAlreadyUsed(proof, state.clone()).into(),
+                    )
+                } else {
+                    ControlFlow::Continue(state)
+                }
+            }
+            None => ControlFlow::Break(ManifestValidationError::ProofNotYetCreated(proof).into()),
+        }
+    }
+
+    fn handle_cloned_proof<V: ManifestInterpretationVisitor>(
+        &mut self,
+        visitor: &mut V,
+        cloned_proof: ManifestProof,
+    ) -> ControlFlow<V::Error<'a>> {
+        let source_amount = self.get_existing_proof::<V>(cloned_proof)?.source_amount;
+        self.handle_new_proof(visitor, source_amount)
+    }
+
+    fn consume_proof<V: ManifestInterpretationVisitor>(
+        &mut self,
+        visitor: &mut V,
+        proof: ManifestProof,
+        destination: ProofDestination,
+    ) -> ControlFlow<V::Error<'a>> {
+        let location = self.location;
+        let proof_state = self.get_existing_proof::<V>(proof)?;
+        proof_state.consumed_at = Some(location);
+        let source_amount = proof_state.source_amount;
+        match source_amount.proof_kind() {
+            ProofKind::BucketProof(bucket) => {
+                self.get_existing_bucket::<V>(bucket)?.proof_locks -= 1;
+            }
+            ProofKind::AuthZoneProof => {}
+        }
+        visitor.on_consume_proof(proof, destination)
+    }
+
+    fn handle_new_address_reservation<V: ManifestInterpretationVisitor>(
+        &mut self,
+        visitor: &mut V,
+        package_address: &PackageAddress,
+        blueprint_name: &str,
+        preallocated_address: Option<&GlobalAddress>,
+    ) -> ControlFlow<V::Error<'a>, ManifestAddressReservation> {
+        let new_address_reservation =
+            ManifestAddressReservation(self.address_reservation_state.len() as u32);
```

Comment:
I think that the logic is currently identical to that of the `ManifestIdAllocator`, so no issues with it being different logic or anything, but why not make use of it here? It seems like it would be better suited for this.

Reply:
Hah well in my head initially I wanted this to replace `ManifestIdAllocator`, but in reality I think id allocator has its niche :).

I'd be comfortable introducing a `ManifestIdAllocator`, but I think on balance I prefer the current code.
By storing in a Vec this implementation requires that the ids be consecutive, so I slightly prefered making it explicit / self-contained in this impl.